src/DayPicker.tsx

```diff
@@ -256,6 +263,118 @@ export function DayPicker(props: DayPickerProps) {
 
   const dataAttributes = getDataAttributes(props);
 
+  const rootRef = useRef<HTMLDivElement>(null);
+  const rootElementSnapshot = useRef<HTMLElement>(null);
+  const previousMonths = useRef(months);
+
+  useLayoutEffect(() => {
+    if (months.length === 0) {
+      return;
+    }
+
+    const isSameMonth = dateLib.isSameMonth(
+      months[0].date,
+      previousMonths.current[0].date
+    );
+    const slideLeft = dateLib.isAfter(
+      months[0].date,
+      previousMonths.current[0].date
+    );
+    previousMonths.current = months;
+    if (rootRef.current) {
+      const previousMonthElement =
+        rootElementSnapshot.current?.querySelector(`[data-month]`);
+      rootElementSnapshot.current = rootRef.current.cloneNode(
+        true
+      ) as HTMLElement;
+      if (isSameMonth) {
+        return;
+      }
+
+      const currentMonthElement = rootRef.current.querySelector(`[data-month]`);
+      if (
+        currentMonthElement &&
+        currentMonthElement instanceof HTMLElement &&
+        previousMonthElement &&
+        previousMonthElement instanceof HTMLElement
+      ) {
+        // animate new displayed month
+        currentMonthElement.style.overflow = "hidden";
+        const monthCaptionElement =
+          rootRef.current.querySelector(`[data-month-caption]`);
+        if (monthCaptionElement && monthCaptionElement instanceof HTMLElement) {
+          monthCaptionElement.style.animation = "fadeIn 0.3s forwards";
+        }
+
+        const weeksElement = rootRef.current.querySelector(`[data-weeks]`);
+        if (weeksElement && weeksElement instanceof HTMLElement) {
+          weeksElement.style.animation = slideLeft
+            ? "slideInRight 0.3s forwards"
+            : "slideInLeft 0.3s forwards";
+        }
+        // animate new displayed month end
+        const cleanUp = () => {
+          if (
+            monthCaptionElement &&
+            monthCaptionElement instanceof HTMLElement
+          ) {
+            monthCaptionElement.style.animation = "";
+          }
+          if (weeksElement && weeksElement instanceof HTMLElement) {
+            weeksElement.style.animation = "";
+          }
+          currentMonthElement.style.overflow = "";
+          if (currentMonthElement.contains(previousMonthElement)) {
+            currentMonthElement.removeChild(previousMonthElement);
+          }
+        };
+
+        // animate old displayed month
+        previousMonthElement.style.pointerEvents = "none";
```

Comment:
Would this cause the focus being "lost"? Consider adding a test case.

Reply:
I checked that `pointer-events` don't affect focus, but fortunately the focus most of the time works fine in this approach, the problem is the focus being applied to the next month when sliding in from the right, for some reason this causes issues to all animations, but semantically the focus is on the correct element when changing month.